general/device.cpp

```diff
@@ -596,4 +596,131 @@ void Device::Setup(const int device_id)
    if (Allows(Backend::DEBUG_DEVICE)) { ngpu = 1; }
 }
 
+MemoryType Device::QueryMemoryType(void* ptr)
+{
+   // from HYPRE's hypre_GetPointerLocation
+   MemoryType res = MemoryType::HOST;
+#if defined(MFEM_USE_CUDA)
+   struct cudaPointerAttributes attr;
+
+#if (CUDART_VERSION >= 10000)
+#if (CUDART_VERSION >= 11000)
+   MFEM_GPU_CHECK(cudaPointerGetAttributes(&attr, ptr));
+#else
+   cudaPointerGetAttributes(&attr, ptr);
+   if (err != cudaSuccess)
+   {
+      /* clear the error */
+      cudaGetLastError();
+   }
+#endif
+   switch (attr.type)
+   {
+      case cudaMemoryTypeUnregistered:
+         // host
+         break;
+      case cudaMemoryTypeHost:
+         res = MemoryType::HOST_PINNED;
+         break;
+      case cudaMemoryTypeDevice:
+         res = MemoryType::DEVICE;
+         break;
+      case cudaMemoryTypeManaged:
+         res = MemoryType::MANAGED;
+         break;
+   }
+#else
```

Comment:
Is this section for CUDA < v10? If so, we don't need it since we require CUDA >= 10.1.168 as specified in the `INSTALL` file.